### ISSUE TITLE
Files pane improvements

### ISSUE THREAD

**Issue by oscargus** [spyder-ide/spyder#15939]
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
I am writing a program that generates lots of results files and I realized I was missing the following features in the Files pane:

 * Force a manual update of the file list (may be related to #10081, not sure how the directory is polled for new files, so cannot tell if it is filling the list that takes time or that an update is not triggered "fast enough"). Use case: I want to check the results file when I see that it has been generated without having to wait for an automatic(?) update of the list.
 * A "dynamic" filter feature similar to the switcher (Ctrl+p) where one can type parts of a file name and only the matching files are shown. Use case: there are lots of results files after a while, so hard to find "old" ones, but I know what the file is called (since it has a filename based on the results).

## Versions

* Spyder version: 5.0.3

**Comment by steff456** [comment]
Hi @oscargus,

Can you please let me know how many files are you generating so I can try to see how much time it is taking to reload on my end?

For your second point, we have a filter feature in the Files pane,
![image](https://user-images.githubusercontent.com/20992645/123664187-bd254980-d7fc-11eb-895b-9545286a16c2.png)

Where you can put any filter you want using a glob expression,
![image](https://user-images.githubusercontent.com/20992645/123664386-eb0a8e00-d7fc-11eb-9d78-8af302e528d6.png)

And then choose if you want to activate/deactivate using the filter push button in toolbar of the pane,
![image](https://user-images.githubusercontent.com/20992645/123664450-f6f65000-d7fc-11eb-9527-d80aba14c3ce.png)

Please let me know if this helps your workflow.

**Comment by oscargus** [comment]
Currently I have about 750 files. I did some more testing and I do not think it is the problem in #10081 that is causing the problem. Navigating to the folder is not really a problem. (It is noticeable that it takes a bit of time, but less than a second, so just so you can see that it is not directly created.) Hence, some manual updating button would make the trick. (Right now, I may have to up one level and down again and then it usually updates.)

I am aware of that filtering functionality. It is indeed nice, but more suitable for "static" filtering, not to find a file on the fly. Hence, I think it should be more useful to have a text box in the toolbar for dynamically filtering the items, similar to the behavior of the switcher. (Plus probably either a check box to enable it or a button to clear the text box and thereby restoring it.)

The Find pane will find the contents of a file, while I am interesting in seeing only certain file names.

I realize that it may be a fringe use case, but whenever you try to load a file from a directory containing many files this should be useful.

(If the fuzzy search that is implemented in the switcher can be used that is of course a bonus, but some simpler variants would also be useful.)

I did look around a bit and it seems like it is not straightforward to force an update of a QFileSystemModel. An option is https://stackoverflow.com/questions/49718552/how-to-update-file-info-in-a-qfilesystemmodel where the file info is watched more carefully. However, it comes at a resource cost, so in that case I think it should be configurable in the preferences if it is used or not (will require a restart).

For the filtering, it seems like one may have to implement a QSortFilterProxyModel to handle it, an example is given at
https://gist.github.com/justinfx/dcdef94b2916ddd5f31e9f0921999fae Not sure if it can be combined with the current filtering or if that has to be reimplemented is the QSortFilterProxyModel as well though.

(Just leaving it here in case someone wants to give it a go.)

**Comment by oscargus** [comment]
> > Hence, some manual updating button would make the trick. (Right now, I may have to up one level and down again and then it usually updates.)
> 
> So what exactly happens for you right now? Are you adding files to a directory but they are not immediately displayed in Files?

Correct. I just arrived at my (Linux) office computer and it still showed the status from yesterday, despite having added about a hundred files overnight. Now, this is an NFS (network) mounted directory, so that may affect the behavior. However, even on my Windows laptop with local drive there is a noticeable delay (not 16 hours though...). (Seems to be a QFileSystemModel-related issue and no straightforward way of solving it, at least not using an "update button", if I understand it correctly.)

Considering the vacations coming up soon and this is quite useful for me at the moment, it may be that I have time to sort it out. That is, adding a PatternComboBox for dynamic filtering, possibly with regex and case sensitivity options (and maybe fuzzy search) and, ideally, adding an option in the preferences to do `os.environ["QT_FILESYSTEMMODEL_WATCH_FILES"] = '1'` when enabled.

Is master or 5.x the correct branch to base the PR on?

Also, in case I go there, is there a preferred icon for fuzzy search? Didn't find anything in the IconManager nor material design icons.

**Comment by ccordoba12** [comment]
> Edit: I read that it is clear that master is the correct branch to base the PR on.

Yeah, that's correct.

> Also, in case I go there, is there a preferred icon for fuzzy search?

The same used in the Variable Explorer to filter variables.
